gap/tools/io.gi

```diff
@@ -313,3 +313,212 @@ function(str)
+  if Length(arg) = 1 then
+    name    := arg[1];
+    line_nr := 0;
+  elif Length(arg) = 2 then
+    name    := arg[1];
+    line_nr := arg[2];
+  else
+    ErrorNoReturn("Semigroups: ReadMultiplicationTable: usage,\n",
+                  "there should be at most 2 arguments,");
+  fi;
+  if IsString(name) then
+    name := UserHomeExpand(name);
+    file := IO_CompressedFile(name, "r");
+    if file = fail then
+      ErrorNoReturn("Semigroups: ReadMultiplicationTable:\n",
+                    "could not open the file \"", name, "\",");
+    fi;
+  elif IsFile(name) then
+    file := name;
+  else
+    ErrorNoReturn("Semigroups: ReadMultiplicationTable: usage,\n",
+                  "the first argument must be a string or a file,");
+  fi;
+  if not (IsInt(line_nr) and line_nr >= 0) then
+    ErrorNoReturn("Semigroups: ReadMultiplicationTable: usage,\n",
+                  "the second argument must be a positive integer,");
+  fi;
+  ReadMultiplicationTableLine := SEMIGROUPS.ReadMultiplicationTableLine;
+
+  if line_nr <> 0 then
+    i := 0;
+    repeat
+      i    := i + 1;
+      line := IO_ReadLine(file);
+    until i = line_nr or line = "";
+    if IsString(arg[1]) then
+      IO_Close(file);
+    elif line = "" then
+      ErrorNoReturn("Semigroups: ReadMultiplicationTable:\n",
+                    "the file only has ", i - 1, " lines,");
+    fi;
+    n := RootInt(Length(line)); #TODO: Returns the int part. Change or check?
+    return ReadMultiplicationTableLine(Chomp(line), n);
+  else
+    lines := IO_ReadLines(file);
+    if IsString(arg[1]) then
+      IO_Close(file);
+    fi;
+    if Length(lines) > 0 then
+      n := RootInt(Length(lines[1])); #TODO: Same as above.
+    else
+      n := 0;
+    fi;
+    Apply(lines, line -> ReadMultiplicationTableLine(Chomp(line), n));
+    return lines;
+  fi;
+end);
+
+SEMIGROUPS.ReadMultiplicationTableLine := function(line, n)
+  local table, vals, row, i, j;
+  table := EmptyPlist(n);
+  vals := [0 .. n - 1];
+  for i in vals do
+    row := EmptyPlist(n);
+    for j in vals do
+      row[j + 1] := IntHexString([line[i * n + j + 1]]) + 1;
+    od;
+    table[i + 1] := row;
+  od;
+  return table;
+end;
+
+InstallGlobalFunction(WriteMultiplicationTable,
+function(arg)
+  local name, coll, mode, file, str, i, n;
+
+  if Length(arg) = 2 then
+    name := arg[1];
+    coll := arg[2];
+    mode := "w";
+  elif Length(arg) = 3 then
+    name := arg[1];
+    coll := arg[2];
+    mode := arg[3];
+  else
+    ErrorNoReturn("Semigroups: WriteMultiplicationTable: usage,\n",
+                  "there should be 2 or 3 arguments,");
+  fi;
+
+  if not (mode = "a" or mode = "w") then
+    ErrorNoReturn("Semigroups: WriteMultiplcationTable: usage,\n",
+                  "the third argument must be \"a\" or \"w\",");
+  fi;
+
+  if IsString(name) then
+    name := UserHomeExpand(name);
+    file := IO_CompressedFile(name, mode);
+    if file = fail then
+      # Cannot test this
```

Comment:
There is an identical comment in the method for WriteGenerators, which isn't covered by tests. What is the difference?